### PR TITLE
opt: directive that asserts a rule was used

### PR DESCRIPTION
**Pull request by justinj**
Brought up in #29050 as a possibility to avoid the case of certain rules
no longer applying.

Kind of a strawman right now because I'm not sure
* if this is the syntax we want,
* how widely we want to use it, I've just added it to one file for the
  purposes of this PR.

I think this is probably valuable in that it asserts we're testing what
we want to test, but I do think it's a little invasive as far as
cluttering up the test directives goes. Open to suggestions for how to
possibly improve.

Release note: None

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
opttester already has a `ruleFromString`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-:-LKwjusO9UqroZ_9WhXV:b-iulybb)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/testutils/opt_tester.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-LKwk67VBQwYnuTGzj-D:-LKwk67VBQwYnuTGzj-E:bjkfgfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/73a0740b593348744a927e1c104a474b6179ccf5/pkg/sql/opt/testutils/opt_tester.go#L96)):*
> ```Go
> 
> 	// DisableRules is a set of rules that are not allowed to run.
> 	DisableRules map[opt.RuleName]struct{}
> ```

Should also be a `RuleSet`

**Review by justinj — COMMENTED** [review]
Oh great - didn't notice that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-:-LKwlg9R0uCaEEbLwBIs:b-qye6xx)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/testutils/opt_tester.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-LKwk67VBQwYnuTGzj-D:-LKwlaDT9dPspo1KOg7j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/73a0740b593348744a927e1c104a474b6179ccf5/pkg/sql/opt/testutils/opt_tester.go#L96)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should also be a `RuleSet`
</blockquote></details>

Done.

**Review by RaduBerinde — COMMENTED** [review]
LGTM.

Note that all flags are documented in the comment for RunCommand, please update it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-:-LKwqL6l9DGMAVUDfIvg:b-mrjxwy)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by rytaft — APPROVED** [review]
:lgtm: I think you can go ahead and add this in other places where we are testing rules - seems useful!

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-:-LKx4SXyAR_BBxLFX6vE:b6wpa78)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by andy-kimball — APPROVED** [review]
:lgtm:, will be interesting to see what issues this turns up

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29135#-:-LKxEnm32FVNMyBEedNO:b-so6r09)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

**Comment by justinj** [comment]
TFTRs! Was going to put the additional directive changes on this PR but after some thought I'm going to split it out and merge this as-is.

bors r+

**Comment by mberhault** [comment]
Bors will never pass this, Lint is failing. I recommend waiting for GitHub CI to pass before asking bors to merge.

**Comment by justinj** [comment]
CI green now

bors r+

**Comment by andreimatei** [comment]
still "wip"?

**Comment by justinj** [comment]
renamed the PR - the commit message was updated before

**Comment by andreimatei** [comment]
This just happens to be one of my frivolous peeves. I think a wip marker
can only hurt; it's frequently forgotten and I need more info somewhere
else (a review comment probably) to know what's in progress anyway.

**Comment by RaduBerinde** [comment]
backport? could be useful for debugging

**Pull request merged**
